# 01 Data Cleaning

## 1.1 Purpose

This notebook turns the standardized transaction file into an analysis-ready product sales table and the supporting SKU tables used by every later notebook.

Input: `data/interim/standardized_sales.csv`, written by `scripts/standardize_raw_sales.py` after `scripts/validate_input_data.py` has checked the raw file against `config/schema_mapping_template.csv`.

The cleaning rules live in `src/retail_analytics/cleaning.py` and are covered by `tests/test_cleaning.py`. This notebook runs them, shows the evidence behind each rule, and saves the outputs.

## 1.2 Cleaning Rules

1. **Normalize types.** `stock_code` is trimmed and upper-cased because the source contains case-only duplicates of the same product (for example `85123a` and `85123A`).
2. **Remove exact duplicate rows.**
3. **Separate returns and cancellations.** Cancellation invoices (`C` prefix) and negative-quantity lines go to `returns_cancellations.csv`.
4. **Keep valid sales lines.** A line needs positive quantity and price and a usable code, description, and date. Missing customer IDs are kept because the analysis is SKU-level.
5. **Exclude non-product lines** listed in `config/non_product_stock_codes.csv`:
   - postage and carriage, including numeric-coded carriage;
   - fees and packing charges;
   - discounts;
   - manual and bad-debt adjustments;
   - samples and gift vouchers;
   - a placeholder line priced at £0.001.

   The list was built by reviewing non-standard stock codes, and descriptions containing words such as VOUCHER, CARRIAGE, and CHARGE.
6. **Remove reviewed manual-credit reversals** listed in `config/manual_reversals.csv`. Some keying errors were reversed with a manual credit (stock code `M`) rather than a cancellation of the same stock code, so the cancellation matching in rule 7 cannot see them. Each listed reversal is checked against the data. `manual_credit_candidates.csv` lists other possible cases for review; nothing is removed from that list automatically.
7. **Remove fully reversed sales.** A sale line is dropped when a later cancellation line from the same customer cancels the same stock code and quantity.
   - Cancellations are processed in time order.
   - Each cancellation takes an unused sale **at the same unit price** when one exists (the most recent of those), otherwise the most recent unused sale.
   - Every match records its price difference and match quality.
   - Partial returns are not netted.
8. **Build the SKU tables.** The SKU master has one row per `stock_code`, with the most frequent description used for display. A monthly SKU sales table is built alongside it.
9. **Flag months truncated by the data's start or end date,** and write a price-anomaly review list (lines priced at least 10 times their SKU's median price).

In [1]:
import sys
from pathlib import Path

project_root = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
if str(project_root / "src") not in sys.path:
    sys.path.insert(0, str(project_root / "src"))

import pandas as pd
from retail_analytics import cleaning

standardized_path = project_root / "data" / "interim" / "standardized_sales.csv"
rules_path = project_root / "config" / "non_product_stock_codes.csv"
manual_reversals_path = project_root / "config" / "manual_reversals.csv"
processed_dir = project_root / "data" / "processed"
processed_dir.mkdir(parents=True, exist_ok=True)

raw = cleaning.load_standardized_sales(standardized_path)
rules = cleaning.load_non_product_rules(rules_path)
manual_reversals = cleaning.load_manual_reversals(manual_reversals_path)

print("Standardized input shape:", raw.shape)
raw.head()

Standardized input shape: (541909, 8)


,invoice_no,stock_code,description,quantity,invoice_date,unit_price,customer_id,country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


## 1.3 Input Profile

A quick profile of the standardized input before any rule is applied.

In [2]:
profile = cleaning.normalize_sales(raw)
print("Missing values:")
print(profile.isna().sum())
print("\nExact duplicate rows:", int(profile.duplicated().sum()))
print("Cancellation invoices:", int(profile["invoice_no"].str.startswith("C", na=False).sum()))
print("Non-positive quantity lines:", int((profile["quantity"] <= 0).sum()))
print("Non-positive price lines:", int((profile["unit_price"] <= 0).sum()))

raw_codes = raw["stock_code"].astype("string").str.strip()
print("\nDistinct stock codes before case normalization:", raw_codes.nunique())
print("Distinct stock codes after case normalization:", raw_codes.str.upper().nunique())

Missing values:
invoice_no           0
stock_code           0
description       1454
quantity             0
invoice_date         0
unit_price           0
customer_id     135080
country              0
dtype: int64



Exact duplicate rows: 5268
Cancellation invoices: 9288
Non-positive quantity lines: 10624
Non-positive price lines: 2517

Distinct stock codes before case normalization: 4070


Distinct stock codes after case normalization: 3958


## 1.4 Rule Tables

Both exclusion lists are configuration, not code, so they can be reviewed and extended without editing the notebook. The loaders reject blank, duplicate, or overlapping non-product rules, and manual reversals that do not match the data.

In [3]:
display(rules)
display(manual_reversals)

,stock_code,match_type,category,reason
0,POST,exact,shipping,Postage charge
1,DOT,exact,shipping,Dotcom postage
2,C2,exact,shipping,Carriage charge
3,23444,exact,shipping,Next day carriage charge (numeric code)
4,BANK CHARGES,exact,fee,Bank charges
5,AMAZONFEE,exact,fee,Marketplace fee
6,23574,exact,fee,Packing charge (numeric code)
7,CRUK,exact,fee,CRUK commission
8,D,exact,adjustment,Discount line
9,M,exact,adjustment,Manual adjustment (also matches lowercase m af...


,invoice_no,stock_code,quantity,unit_price,credit_invoice_no,reason
0,556444,22502,60,649.5,C556445,60 x 649.50 keyed in error (line value 38970);...


## 1.5 Apply the Cleaning Rules

In [4]:
result = cleaning.clean_transactions(raw, rules, manual_reversals)
sales = result.sales
returns_cancellations = result.returns_cancellations
non_product_lines = result.non_product_lines
reversed_sales = result.reversed_sales
manual_credit_candidates = result.manual_credit_candidates

pd.Series(result.step_counts, name="row_count").to_frame()

,row_count
standardized_input,541909
after_duplicate_removal,536641
returns_cancellations_separated,10587
valid_sales_before_exclusions,524878
non_product_lines_excluded,2435
manual_credit_reversals_removed,1
fully_reversed_sales_removed,2815
valid_product_sales_final,519627


## 1.6 Evidence: Non-Product Lines Excluded

In [5]:
non_product_summary = (
    non_product_lines
    .assign(line_value=lambda d: d["quantity"] * d["unit_price"])
    .groupby(["non_product_category", "stock_code"], as_index=False)
    .agg(rows=("invoice_no", "size"), line_value=("line_value", "sum"))
    .sort_values("line_value", ascending=False)
)
non_product_summary

,non_product_category,stock_code,rows,line_value
15,shipping,DOT,706,206248.770
16,shipping,POST,1126,78101.880
1,adjustment,M,317,77752.820
3,fee,AMAZONFEE,2,13761.090
0,adjustment,B,1,11062.060
14,shipping,C2,141,7051.000
13,shipping,23444,79,1200.000
8,non_merchandise,GIFT_0001_30,7,175.530
10,non_merchandise,GIFT_0001_50,4,167.560
7,non_merchandise,GIFT_0001_20,9,167.050


## 1.7 Evidence: Reversed Sales

The two largest lines in the raw data were each cancelled within minutes:
- 80,995 units of `23843`;
- 74,215 units of `23166`.

Keeping them would overstate revenue and make both SKUs look like top sellers.

Customer 15098 on 10 June 2011 shows why both matching rules are needed:

| Invoice | Line | What happened |
| --- | --- | --- |
| 556442 | 60 × £4.95 | Cancelled by C556448 at the same price |
| 556444 | 60 × £649.50 = £38,970 | Keying error, reversed by manual credit C556445 for £38,970 |
| 556446 | 1 × £649.50 | The corrected invoice, which is kept |

Matching only on customer, code, and quantity would have paired C556448 with 556444, because 556444 is the more recent sale.

In [6]:
summary_by_type = (
    reversed_sales.groupby("reversal_type")
    .agg(lines=("invoice_no", "size"), revenue_removed=("revenue", "sum"))
    .round(2)
)
display(summary_by_type)
display(reversed_sales["match_quality"].value_counts().rename("lines").to_frame())

full = reversed_sales[reversed_sales["reversal_type"] == "full_cancellation"]
same_day_share = full.loc[full["days_to_cancel"] == 0, "revenue"].sum() / full["revenue"].sum()
print(f"Share of cancellation-matched revenue cancelled on the same day: {same_day_share:.1%}")

reversed_sales.nlargest(10, "revenue")[
    ["invoice_no", "cancel_invoice_no", "reversal_type", "match_quality", "stock_code",
     "quantity", "unit_price", "cancel_unit_price", "price_delta", "revenue", "days_to_cancel"]
]

,lines,revenue_removed
reversal_type,,
full_cancellation,2815,387988.81
manual_credit,1,38970.00


,lines
match_quality,
same_price,2722
price_mismatch,93
reviewed_manual_credit,1


Share of cancellation-matched revenue cancelled on the same day: 75.4%


,invoice_no,cancel_invoice_no,reversal_type,match_quality,stock_code,quantity,unit_price,cancel_unit_price,price_delta,revenue,days_to_cancel
2815,581483,C581484,full_cancellation,same_price,23843,80995,2.08,2.08,0.0,168469.60,0
186,541431,C541433,full_cancellation,same_price,23166,74215,1.04,1.04,0.0,77183.60,0
0,556444,C556445,manual_credit,reviewed_manual_credit,22502,60,649.50,38970.00,NaN,38970.00,0
734,540815,C550456,full_cancellation,same_price,21108,3114,2.10,2.10,0.0,6539.40,97
732,540815,C550456,full_cancellation,same_price,85123A,1930,2.55,2.55,0.0,4921.50,97
731,540818,C550456,full_cancellation,same_price,48185,670,6.75,6.75,0.0,4522.50,97
1727,567423,C567527,full_cancellation,same_price,23113,756,5.06,5.06,0.0,3825.36,0
733,540815,C550456,full_cancellation,same_price,21175,2000,1.85,1.85,0.0,3700.00,97
2033,570094,C570556,full_cancellation,same_price,22273,720,2.55,2.55,0.0,1836.00,3
2028,570097,C570556,full_cancellation,same_price,22273,720,2.30,2.30,0.0,1656.00,3


### Manual-credit candidates (review only)

Manual credits whose value equals a sale line from the same customer within the previous day. Only reviewed entries in `config/manual_reversals.csv` are removed.

In [7]:
manual_credit_candidates

,credit_invoice_no,credit_date,customer_id,credit_amount,candidate_count,candidate_invoices,candidate_stock_codes,reviewed_in_config
0,C556445,2011-06-10 15:31:00,15098,38970.00,1,556444,22502,True
1,C578073,2011-11-22 16:02:00,18139,20.16,2,577783;577816,22712;22712,False
2,C546870,2011-03-17 16:07:00,14462,5.00,2,546866;546866,77101A;21390,False
3,C567867,2011-09-22 14:18:00,16669,4.00,1,567866,16045,False
4,C558712,2011-07-01 13:06:00,17338,2.95,3,558709;558709;558709,22328;22326;22667,False


## 1.8 SKU Master and Description Check

`stock_code` is the SKU key. `description_count` shows how many descriptions one code carries; the most frequent description is used for display.

In [8]:
sku_master = cleaning.build_sku_master(sales)
sku_description_check = (
    sku_master[["stock_code", "description", "description_count", "total_units", "total_revenue"]]
    .sort_values("description_count", ascending=False)
)

print("SKU master shape:", sku_master.shape)
print("SKUs with more than one description:", int((sku_master["description_count"] > 1).sum()))
sku_description_check.head(10)

SKU master shape: (3790, 8)
SKUs with more than one description: 209


,stock_code,description,description_count,total_units,total_revenue
2041,23196,VINTAGE LEAF MAGNETIC NOTEPAD,4,1872,2842.24
2081,23236,STORAGE TIN VINTAGE DOILY,4,2486,7039.74
2220,23396,LE JARDIN BOTANIQUE CUSHION COVER,3,650,2735.53
2355,23535,WALL ART BICYCLE SAFETY,3,1323,8434.29
2048,23203,JUMBO BAG VINTAGE DOILY,3,20144,41413.86
2054,23209,LUNCH BAG VINTAGE DOILY,3,13149,21384.17
2076,23231,WRAP DOILEY DESIGN,3,7277,3008.34
2085,23240,SET OF 4 KNICK KNACK TINS DOILY,3,4164,16946.22
2089,23244,ROUND STORAGE TIN VINTAGE LEAF,3,935,1732.03
100,17107D,"FLOWER FAIRY,5 SUMMER B'DRAW LINERS",3,170,433.50


## 1.9 Monthly SKU Sales, Month Coverage, and Price Review

The data ends on 9 December 2011, so December 2011 is flagged as a truncated month; trend views should not compare it with full months. The flag only covers the first and last month of the data.

The price-anomaly list flags lines priced at least 10 times their SKU's median price. It is a review aid; no line is removed.

In [9]:
monthly_sku_sales = cleaning.build_monthly_sku_sales(sales, sku_master)
month_coverage = cleaning.month_coverage(sales["invoice_date"])
price_anomalies = cleaning.find_price_anomalies(sales)

print("Monthly SKU sales shape:", monthly_sku_sales.shape)
display(month_coverage)

print(f"Price-anomaly lines: {len(price_anomalies):,}")
print(f"Total value of flagged lines: {price_anomalies['line_value'].sum():,.2f}")
print(f"Share without customer ID: {price_anomalies['customer_id'].isna().mean():.0%}")
price_anomalies.head(10)

Monthly SKU sales shape: (33360, 7)


,invoice_month,first_date,last_date,is_truncated_month
0,2010-12,2010-12-01,2010-12-23,False
1,2011-01,2011-01-04,2011-01-31,False
2,2011-02,2011-02-01,2011-02-28,False
3,2011-03,2011-03-01,2011-03-31,False
4,2011-04,2011-04-01,2011-04-28,False
5,2011-05,2011-05-01,2011-05-31,False
6,2011-06,2011-06-01,2011-06-30,False
7,2011-07,2011-07-01,2011-07-31,False
8,2011-08,2011-08-01,2011-08-31,False
9,2011-09,2011-09-01,2011-09-30,False


Price-anomaly lines: 387
Total value of flagged lines: 4,963.05
Share without customer ID: 99%


,invoice_no,stock_code,description,quantity,unit_price,sku_median_price,price_ratio,line_value,invoice_date,customer_id
0,556446,22502,PICNIC BASKET WICKER 60 PIECES,1,649.50,5.95,109.2,649.50,2011-06-10 15:33:00,15098
1,538071,22945,CHRISTMAS METAL TAGS ASSORTED,6,11.02,0.85,13.0,66.12,2010-12-09 14:09:00,<NA>
2,538566,22945,CHRISTMAS METAL TAGS ASSORTED,5,11.02,0.85,13.0,55.10,2010-12-13 11:21:00,<NA>
3,539451,37370,RETRO COFFEE MUGS ASSORTED,3,16.13,1.25,12.9,48.39,2010-12-17 16:59:00,<NA>
4,538148,37370,RETRO COFFEE MUGS ASSORTED,3,16.13,1.25,12.9,48.39,2010-12-09 16:26:00,<NA>
5,538883,37370,RETRO COFFEE MUGS ASSORTED,3,16.13,1.25,12.9,48.39,2010-12-14 15:56:00,<NA>
6,539434,22945,CHRISTMAS METAL TAGS ASSORTED,4,11.02,0.85,13.0,44.08,2010-12-17 14:41:00,<NA>
7,560090,40016,CHINESE DRAGON PAPER LANTERNS,5,8.29,0.42,19.7,41.45,2011-07-14 16:46:00,<NA>
8,537237,22945,CHRISTMAS METAL TAGS ASSORTED,3,11.02,0.85,13.0,33.06,2010-12-06 09:58:00,<NA>
9,537434,22945,CHRISTMAS METAL TAGS ASSORTED,3,11.02,0.85,13.0,33.06,2010-12-06 16:57:00,<NA>


## 1.10 Data Quality Summary and Saved Outputs

In [10]:
step_notes = {
    "standardized_input": "Rows in data/interim/standardized_sales.csv",
    "after_duplicate_removal": "Rows after removing exact duplicates",
    "returns_cancellations_separated": "Cancellation invoices and negative-quantity lines saved separately",
    "valid_sales_before_exclusions": "Valid sales lines before non-product and reversal exclusions",
    "non_product_lines_excluded": "Lines matched by config/non_product_stock_codes.csv",
    "manual_credit_reversals_removed": "Sales reversed by a reviewed manual credit (config/manual_reversals.csv)",
    "fully_reversed_sales_removed": "Sales fully cancelled later by the same customer (same price preferred)",
    "valid_product_sales_final": "Final product sales lines used for analysis",
}
data_quality_summary = pd.DataFrame(
    [(step, count, step_notes[step]) for step, count in result.step_counts.items()],
    columns=["step", "row_count", "note"],
)
extra = pd.DataFrame([
    ("monthly_sku_sales_records", len(monthly_sku_sales), "Rows in the monthly SKU sales table"),
    ("sku_master_records", len(sku_master), "Distinct normalized stock codes"),
], columns=["step", "row_count", "note"])
data_quality_summary = pd.concat([data_quality_summary, extra], ignore_index=True)

outputs = {
    "clean_sales.csv": sales,
    "returns_cancellations.csv": returns_cancellations,
    "non_product_lines.csv": non_product_lines,
    "reversed_sales.csv": reversed_sales,
    "sku_master.csv": sku_master,
    "sku_description_check.csv": sku_description_check,
    "monthly_sku_sales.csv": monthly_sku_sales,
    "month_coverage.csv": month_coverage,
    "manual_credit_candidates.csv": manual_credit_candidates,
    "price_anomalies.csv": price_anomalies,
    "data_quality_summary.csv": data_quality_summary,
}
for filename, table in outputs.items():
    table.to_csv(processed_dir / filename, index=False)

data_quality_summary

,step,row_count,note
0,standardized_input,541909,Rows in data/interim/standardized_sales.csv
1,after_duplicate_removal,536641,Rows after removing exact duplicates
2,returns_cancellations_separated,10587,Cancellation invoices and negative-quantity li...
3,valid_sales_before_exclusions,524878,Valid sales lines before non-product and rever...
4,non_product_lines_excluded,2435,Lines matched by config/non_product_stock_code...
5,manual_credit_reversals_removed,1,Sales reversed by a reviewed manual credit (co...
6,fully_reversed_sales_removed,2815,Sales fully cancelled later by the same custom...
7,valid_product_sales_final,519627,Final product sales lines used for analysis
8,monthly_sku_sales_records,33360,Rows in the monthly SKU sales table
9,sku_master_records,3790,Distinct normalized stock codes


## 1.11 Summary

In [11]:
print("===== 01 Data Cleaning Summary =====")
print(f"Valid product sales lines: {len(sales):,}")
print(f"Revenue: {sales['revenue'].sum():,.2f}")
print(f"Orders: {sales['invoice_no'].nunique():,}")
print(f"SKUs: {len(sku_master):,}")
print(f"Date range: {sales['invoice_date'].min()} to {sales['invoice_date'].max()}")
print("\nSaved to data/processed/:")
for filename in outputs:
    print("-", filename)

===== 01 Data Cleaning Summary =====
Valid product sales lines: 519,627
Revenue: 9,818,872.18
Orders: 19,646
SKUs: 3,790
Date range: 2010-12-01 08:26:00 to 2011-12-09 12:50:00

Saved to data/processed/:
- clean_sales.csv
- returns_cancellations.csv
- non_product_lines.csv
- reversed_sales.csv
- sku_master.csv
- sku_description_check.csv
- monthly_sku_sales.csv
- month_coverage.csv
- manual_credit_candidates.csv
- price_anomalies.csv
- data_quality_summary.csv
